windows-gaming-driver/src/setup/usb.rs

```diff
@@ -103,51 +201,56 @@ fn pick(special: Option<HidKind>,
     let mut devs = Vec::new();
 
     for dev in iter.scan_devices()? {
-        let dev = if special.is_some() {
+        let dev = if kind.is_some() {
             dev.parent().unwrap()
         } else {
             dev
         };
 
-        let mut vendor = None;
-        let mut product = None;
-        let mut vendor_name = "Unknown vendor".to_owned();
-        let mut product_name = "Unknown product".to_owned();
-
-        for attr in dev.properties() {
-            if let Some(val) = attr.value().to_str() {
-                if attr.name() == "ID_VENDOR_ID" {
-                    vendor = parse_hex(val);
-                } else if attr.name() == "ID_MODEL_ID" {
-                    product = parse_hex(val);
-                } else if attr.name() == "ID_VENDOR_FROM_DATABASE" {
-                    vendor_name = val.to_owned();
-                } else if attr.name() == "ID_MODEL_FROM_DATABASE" {
-                    product_name = val.to_owned();
-                }
-            }
-        }
-
-        let id = (vendor.unwrap(), product.unwrap()).into();
-        if !blacklist.contains(&UsbBinding::ById(id)) {
-            println!("[{}]\t{} {} [{:04x}:{:04x}]", devs.len(), vendor_name, product_name,
-                     id.vendor, id.product);
-            devs.push(Some(id));
-        }
-    }
+        let vendor = dev.property_value("ID_VENDOR_ID")
+            .and_then(|s| s.to_str())
+            .and_then(parse_hex)
+            .unwrap();
+        let product = dev.property_value("ID_MODEL_ID")
+            .and_then(|s| s.to_str())
+            .and_then(parse_hex)
+            .unwrap();
+        let vendor_name = dev.property_value("ID_VENDOR_FROM_DATABASE")
+            .map(|s| s.to_string_lossy().into_owned());
+        let product_name = dev.property_value("ID_MODEL_FROM_DATABASE")
+            .map(|s| s.to_string_lossy().into_owned());
+        let busnum = dev.attribute_value("busnum")
+            .and_then(|s| s.to_str())
+            .and_then(parse_hex)
+            .unwrap();
+        let devpath = dev.attribute_value("devpath")
+            .and_then(|s| s.to_str())
+            .and_then(parse_hex)
+            .unwrap();
```

Comment:
Now this is just overly verbose (L210-L229). You have `to_str`+`parse_hex` 4 times, why not move it into `parse_hex` as I suggested? At least create a nested helper function (or even lambda).